### ISSUE TITLE
Heredoc highlighting

### ISSUE THREAD

**Issue by dasisdormax** [jfcherng/Sublime-ShellScriptImproved#1]
First of all, huge respect and thanks for writing this package. (It's probably really difficult).

I found that you can break the highlighting quite hard using some heredocs (but the highlighting included in sublime text has it's issues too ^^)

Anyway, here's the code you can test it on. Tested on build 3103, if it helps.

[heredocs.sh.txt](https://github.com/jfcherng/Sublime-ShellScriptImproved/files/206021/heredocs.sh.txt) (github somehow doesn't allow to upload .sh files)

Honestly, other termination strings aren't too important, quite everyone uses EOF

**Comment by jfcherng** [comment]
Should be resolved by https://github.com/jfcherng/Sublime-ShellScriptImproved/commit/9012a78b347066a3ded4c2f1d3c0aea8213c0984.

![image](https://cloud.githubusercontent.com/assets/6594915/14312058/218a5936-fc1d-11e5-81a8-8a91c59250e2.png)

## 

I am not sure whether to highlight a customized heredoc. iirc, I remove it from the official ST's syntax definition...

**Comment by dasisdormax** [comment]
Yes, that fix is fine for me this way, thanks. I don't specifically need customized heredocs, just pointed it out.

However, the highlighting will still stop working if you have enough code in a file like the following ... (try copy-pasting multiple times)

``` bash
if cond; then cat <<EOF; do_cleanup; exit 1; fi
Error: Condition is met.
EOF

# OR

print_info_text () { cat <<EOF; }
This is some info text
This is another line
EOF
```

It seems the fi or closing bracket after the heredoc start aren't really registered and the if-then block and function aren't properly closed (while they are still properly highlighted).

I don't know though, if this is something that you can fix (with reasonable effort, of course) or if this is an issue of sublime text itself.

If you don't think, it is worth the effort, I can still put the fi on the line after the EOF. I don't mean to cause you headaches with this :)

**Comment by jfcherng** [comment]
Oh! I did not notice that. Indeed, the scope after your test cases are not correct. It just looks like `fi` and `}` do not exist. In fact, when the parser meets `<<EOF`, the context goes to heredoc hence the `fi` is not properly matched to the `if` since the `if` is not in the current context.

As far as my understanding about how Sublime Text's syntax parsing works, currently, there may be no way to know previous context. I would check this with other developers. Thanks for reporting this. :+1: 

## 

UPDATE: Scope issues should be resolved in 1.0.12.
